rtrlib/pfx/trie/trie-pfx.c

```diff
@@ -0,0 +1,472 @@
+        return PFX_SUCCESS;
+    }
+
+    unsigned int lvl = 0;
+    struct trie_node *node = trie_lookup_first_match(root, prefix, prefix_len, &lvl);
+    if(node == NULL) {
+        pthread_rwlock_unlock(&pfx_table->lock);
+        *result = BGP_PFXV_STATE_NOT_FOUND;
+        pfx_table_free_reason(reason, reason_len);
+        return PFX_SUCCESS;
+    }
+
+    if(reason_len != NULL && reason != NULL) {
+        *reason_len = ((struct node_data *) node->data)->len;
+        *reason = realloc(*reason, *reason_len * sizeof(struct pfx_record));
+        if(*reason == NULL) {
+            pthread_rwlock_unlock(&pfx_table->lock);
+            pfx_table_free_reason(reason, reason_len);
+            return PFX_ERROR;
+        }
+        if(pfx_table_node2pfx_record(node, *reason, *reason_len) == PFX_ERROR) {
+            pthread_rwlock_unlock(&pfx_table->lock);
+            pfx_table_free_reason(reason, reason_len);
+            return PFX_ERROR;
+        }
+    }
+
+    while(!pfx_table_elem_matches(node->data, asn, prefix_len)) {
+        if(lrtr_ip_addr_is_zero(lrtr_ip_addr_get_bits(prefix, lvl++, 1))) //post-incr lvl, trie_lookup is performed on child_nodes => parent lvl + 1
+            node = trie_lookup_first_match(node->lchild, prefix, prefix_len, &lvl);
+        else
+            node = trie_lookup_first_match(node->rchild, prefix, prefix_len, &lvl);
+
+        if(node == NULL) {
+            pthread_rwlock_unlock(&pfx_table->lock);
+            *result = BGP_PFXV_STATE_INVALID;
+            return PFX_SUCCESS;
+        }
+
+        if(reason_len != NULL && reason != NULL) {
+            unsigned int r_len_old = *reason_len;
+            *reason_len += ((struct node_data *) node->data)->len;
+            *reason = realloc(*reason, *reason_len * sizeof(struct pfx_record));
+            struct pfx_record *start = *reason + r_len_old;
+            if(*reason == NULL) {
+                pthread_rwlock_unlock(&pfx_table->lock);
+                pfx_table_free_reason(reason, reason_len);
+                return PFX_ERROR;
+            }
+            if(pfx_table_node2pfx_record(node, start, ((struct node_data *) node->data)->len) == PFX_ERROR) {
+                pthread_rwlock_unlock(&pfx_table->lock);
+                pfx_table_free_reason(reason, reason_len);
+                return PFX_ERROR;
+            }
+        }
+    }
+
+    pthread_rwlock_unlock(&pfx_table->lock);
+    *result = BGP_PFXV_STATE_VALID;
+    return PFX_SUCCESS;
+}
+
+int pfx_table_validate(struct pfx_table *pfx_table, const uint32_t asn, const struct lrtr_ip_addr *prefix, const uint8_t prefix_len, enum pfxv_state *result)
+{
+    return pfx_table_validate_r(pfx_table, NULL, NULL, asn, prefix, prefix_len, result);
+}
+
+int pfx_table_src_remove(struct pfx_table *pfx_table, const struct rtr_socket *socket)
+{
+    for(unsigned int i = 0; i < 2; i++) {
+        struct trie_node **root = (i == 0 ? &(pfx_table->ipv4) : &(pfx_table->ipv6));
```

Comment:
The pointer ```pfx_table->ipvX``` will point to an invalid memory address if someone removes the last node form ```pfx_table->ipvX``` and ```pfx_table->ipvX``` gets freed because it was the last node.

So, you should get the R/W lock before getting the reference.